TO: Add 'Geolimit Redirect URL' parameter in Delivery service

xiaodzha opened the pull request:
Codes for #1518

dneuman64 commented:
Can you please fix the conflicts?

dneuman64 commented:
I ran the tests and it looks like a couple failed, can you please fix?  

> Test Summary Report
> -------------------
> t/api/1.0/data.t                        (Wstat: 512 Tests: 88 Failed: 2)
>   Failed tests:  9, 11
>   Non-zero exit status: 2
> t/deliveryservice.t                     (Wstat: 11264 Tests: 72 Failed: 44)
>   Failed tests:  8, 11, 14, 16, 18, 20, 22, 25-33, 35, 37
>                 40-64, 67
>   Non-zero exit status: 44

dneuman64 commented on traffic_ops/app/templates/delivery_service/add.html.ep:
```diff
@@ -162,9 +180,9 @@
-            $('#regional_geo_blocking_row').show(speed);
             $('#regexp_add').show(speed);
             $('#geo_limit_countries_row').show(speed);
+            $('#geolimit_redirect_url_row').hide(speed);
```
should this use your check_geolimit_redirect_url_row_showhide function?

dneuman64 commented:
When I tried to edit an existing DS I get this error 

>  Invalid parameter 'ds.geolimit_redirect_url': cannot use 'geolimit_redirect_url' on a Schema::Result::Deliveryservice at template delivery_service/_form.html.ep line 209.

Can you run `db/admin.pl reverse_schema` from traffic_ops/app to update the schema file?   You will then need to add the schema file to your PR.

xiaodzha commented:
Hi David,

Sorry for the late updates.  I have fixed all the issues you mentioned above.   By the way, after I ran command: db/admin.pl reverse_schema ,  there were some modifications in schema file which was not related to 'geolimit_redirect_url' param, so I removed it manually.  Please let me know if there is anything else I need to update.  

Thanks!
Leo Z

dneuman64 commented:
Great, thanks!  I will take a look when I get a chance.

dneuman64 commented:
A few small things: 
1. Can you update the traffic ops documents to include the new field?  http://traffic-control-cdn.net/docs/latest/admin/traffic_ops_using.html#delivery-service
2. Can you update the API and docs to include the new field?  https://github.com/Comcast/traffic_control/blob/master/traffic_ops/app/lib/API/DeliveryService.pm
3. If I put an invalid URL into the field (for example:  thisisnotavalidurl), it accepts it.  Can you double check your validation?  https://github.com/Comcast/traffic_control/pull/1519/files#diff-dd2f89cbcb46d2fa1240f754fb07682bL523

xiaodzha commented:
Hi @dneuman64 ,
 
I have committed codes for the issues 1 & 2.     As for issue 3,  it should be legal to input that kind of string in the field, for more detailed information, please refer to: http://traffic-control-cdn.net/docs/latest/admin/traffic_router.html#geolimit-failure-redirect-feature

dneuman64 commented:
@xiaodzha thanks for making the changes.
👍
